### ISSUE TITLE
Refactor hideable columns

### ISSUE THREAD

**Issue by gnomeontherun** [vmware/clarity#1856]
**Select one ...**  (check one with "x")
```
[ ] bug
[ ] feature request
[x] enhancement
```

### Expected behavior
Our current API for hideable columns is not as clean as we'd like. Eudes and I talked about the following as the desired API.

```
<clr-datagrid>
    <clr-dg-column *clrIfActive>
        User ID
    </clr-dg-column>
    <clr-dg-column *clrIfActive>
        Name
    </clr-dg-column>
    <clr-dg-row *clrDgItems="let user of users" [clrDgItem]="user">
        <clr-dg-cell *clrIfActive>{{user.id}}</clr-dg-cell>
        <clr-dg-cell>{{user.name}}</clr-dg-cell>
        ...
    </clr-dg-row>
    <clr-dg-footer>
        <clr-dg-column-picker>
            <!-- Project i18n buttons and title here -->
            <clr-dg-column-picker-button [type]="'ok'">OK</clr-dg-column-picker-button>
            <clr-dg-column-picker-button [type]="'selectAll'">Select All</clr-dg-column-picker-button>
            <clr-dg-column-picker-title>Select columns</clr-dg-column-picker-title>
        </clrdg-column-picker>
    </clr-dg-footer>
</clr-datagrid>
```

* Names can be tweaked for clarity as necessary
* This makes it easier to write and manage the lifecycle of the columns
* The individual cells in the row can be optionally removed from the DOM using `clrIfActive` directive just like the columns, otherwise they get `display: none` like they do today
* This is a breaking change
* The presence of the `clr-dg-column-picker` enables or disables the feature
* The presence of the `clrIfActive` directive on the column is what enables a particular column for toggling (can be initialized `*clrIfActive="false"` to disable by default)
* Main challenge here is the TemplateRef contains the actual column when we just want the text from it, need a clean way to extract this

### Reproduction of behavior

* Include a link to the reproduction scenario you created by forking one of the Clarity Plunker Templates:

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:**

* **OS and version:** 

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

**Comment by qbix0r** [comment]
While we talk about hideable columns, there is one thing that bugs me. I want to save the state of visible / hidden columns, so I need to capture the visibility change event. What I do:

```typescript
    public ngAfterViewInit() {
        this.datagridFooter.hideableColumnService.getColumns().forEach((col) => {
            col.hiddenChangeState.subscribe((hidden) => {
                let colState = {};
                this.datagridFooter.hideableColumnService.getColumns().forEach((v) => {
                    colState[v.id] = v.hidden;
                });

                this.hiddenColumns = colState;
                localStorage.setItem('ordersColumns', JSON.stringify(colState));
            });
        });
    }
```
Everything works great, but it looks that everytime a new table is created somewhere in app, every column gets the id property incremented `dg-col-X`.
Is it possible to provide my own column ID or keep separate counter for each table?
This doesn't work
```html
            <clr-dg-column id="idColumn" #idColumn style="width: 150px;" [clrDgSortBy]="'id'">
                <ng-container *clrDgHideableColumn="{hidden: hiddenColumns ? hiddenColumns[idColumn.columnId] : false}">
                    {{'ORDERS.ID' | translate}}
                </ng-container>
            </clr-dg-column>
```

If my approach is wrong, how would you solve it differently?

**Comment by gnomeontherun** [comment]
@qbix0r This is a good question. I think we'll still need to generate a unique ID for each row, but we could allow you to pass a custom identifier that we send back so you can link columns. Thanks for adding that idea to this.

**Comment by Kevin2iBi** [comment]
Any update about this request???

**Comment by mathisscott** [comment]
@gnomeontherun 
Could you assign this to one of our milestones?

**Comment by mathisscott** [comment]
Part of PR #3240

**Comment by gnomeontherun** [comment]
Closed by #3240
